README.rst
|Build Status| |Coverage Status| |PyPI|

ps_alchemy
==========

`ps_alchemy` is extension for `pyramid_sacrud
<https://github.com/sacrud/pyramid_sacrud/>`_ which provides SQLAlchemy models.

Look how easy it is to use:

.. code-block:: python

   Base = declarative_base()
   DBSession = scoped_session(sessionmaker(extension=ZopeTransactionExtension()))

   class PageTree(Base):
       __tablename__ = 'pages'

       id = Column(Integer, primary_key=True)

.. code-block:: python

   config.include('ps_alchemy')
   config.include('pyramid_sacrud', route_prefix='admin')
   config.registry.settings['pyramid_sacrud.models'] = ('', PageTree)

For more docs see http://ps-alchemy.rtfd.org

Support and Development
=======================

To report bugs, use the `issue tracker
<https://github.com/sacrud/ps_alchemy/issues>`_

We welcome any contribution: suggestions, ideas, commits with new futures,
bug fixes, refactoring, docs, tests, translations etc

If you have question, contact me sacrud@uralbash.ru or IRC channel #sacrud

License
=======

The project is licensed under the MIT license.

.. |Build Status| image:: https://travis-ci.org/sacrud/ps_alchemy.svg?branch=master
   :target: https://travis-ci.org/sacrud/ps_alchemy
.. |Coverage Status| image:: https://coveralls.io/repos/sacrud/ps_alchemy/badge.png?branch=master
   :target: https://coveralls.io/r/sacrud/ps_alchemy?branch=master
.. |PyPI| image:: http://img.shields.io/pypi/dm/ps_alchemy.svg
   :target: https://pypi.python.org/pypi/ps_alchemy/



CHANGES.rst
0.0.7 (2016-05-07)
==================

- fix UnicodeDecode error (see #6 #7)
- added ``items_per_page`` option to settings (issue
  sacrud/pyramid_sacrud#120)
- fix infinite database connections (issue sacrud/pyramid_sacrud#121)

0.0.6 (2016-04-27)
==================

- fix datetime widget (issue sacrud/pyramid_sacrud#119)

0.0.5 (2016-03-15)
==================

- migrate to ps_crud module

0.0.4 (2016-02-07)
==================

- migrate to pyramid_sacrud == 0.3.2

0.0.3 (2016-01-08)
==================

- clean requirements

0.0.2 (2016-01-08)
==================

- add template folder to MANIFEST (issue #3)

0.0.1 (2016-01-07)
==================

- Hi!


tests/conftest.py
import imp
import os

import pytest
from pyramid import testing
from webtest import TestApp
from sqlalchemy import create_engine
from pyramid_sacrud import CONFIG_RESOURCES
from sqlalchemy.orm import sessionmaker, scoped_session
from pyramid.threadlocal import get_current_registry
from sqlalchemy.ext.declarative import declarative_base

imp.load_source('pyramid_sacrud_example', 'example/ps_alchemy_example.py')

from pyramid_sacrud_example import (  # noqa
    Good,
    main,
    Tree,
    User,
    Group,
    models
)

DIRNAME = os.path.dirname(__file__)
DATABASE_FILE = os.path.join(DIRNAME, 'test.sqlite')
TEST_DATABASE_CONNECTION_STRING = 'sqlite:///%s' % DATABASE_FILE

Base = declarative_base()


class BaseFixture:

    def __init__(self):
        engine = create_engine(TEST_DATABASE_CONNECTION_STRING)
        self.engine = engine
        self.DBSession = scoped_session(sessionmaker(bind=self.engine))
        self.dbsession = self.DBSession(bind=self.engine)

    def setUp(self):
        # bind an individual Session to the connection
        self.create_db()

    def tearDown(self):
        self.drop_db()
        os.remove(DATABASE_FILE)
        self.dbsession.close()

    def drop_db(self):
        Base.metadata.drop_all(bind=self.engine)
        User.__table__.drop(self.engine)
        Group.__table__.drop(self.engine)
        Good.__table__.drop(self.engine)
        Tree.__table__.drop(self.engine)
        self.dbsession.commit()

    def create_db(self):
        Base.metadata.create_all(bind=self.engine)
        User.__table__.create(self.engine, checkfirst=True)
        Group.__table__.create(self.engine, checkfirst=True)
        Good.__table__.create(self.engine, checkfirst=True)
        Tree.__table__.create(self.engine, checkfirst=True)
        self.dbsession.commit()


@pytest.fixture(scope="function")
def unit(request):
    response = BaseFixture()
    response.setUp()
    response.request = testing.DummyRequest()
    response.config = testing.setUp(request=response.request)
    get_current_registry().settings[CONFIG_RESOURCES] = models
    get_current_registry().settings['sqlalchemy.url'] =\
        TEST_DATABASE_CONNECTION_STRING

    def fin():
        response.tearDown()
        testing.tearDown()

    request.addfinalizer(fin)
    return response


@pytest.fixture(scope="function")
def functional(request):
    settings = {
        'sqlalchemy.url': TEST_DATABASE_CONNECTION_STRING,
    }
    app = main({}, **settings)
    response = BaseFixture()
    response.setUp()
    response.app = TestApp(app)

    def fin():
        response.tearDown()

    request.addfinalizer(fin)
    return response


tests/test_views.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2014 uralbash <root@uralbash.ru>
#
# Distributed under terms of the MIT license.

"""
Tests of views
"""
import imp
from collections import OrderedDict

from ps_alchemy.views import Read, Create
from ps_alchemy.resources import (
    ListResource,
    CreateResource,
    UpdateResource
)
from pyramid.httpexceptions import HTTPFound

from .conftest import User

imp.load_source('pyramid_sacrud_example', 'example/ps_alchemy_example.py')

from pyramid_sacrud_example import add_fixtures  # noqa


class TestList:

    def test_empty_user_table(self, unit):
        context = ListResource(User)
        context.renderer = 'json'
        response = Read(context, unit.request).list_view()
        assert response.json['paginator'] == []

    def test_non_empty_user_table(self, unit):
        add_fixtures(unit.dbsession)
        context = ListResource(User)
        context.renderer = 'json'
        response = Read(context, unit.request).list_view()
        assert response.json['paginator'] == [
            {'name': 'admin'},
            {'name': 'moderator'},
            {'name': 'user1'},
            {'name': 'user2'}
        ]


class TestCreate(object):

    def test_create_get_form(self, unit):
        context = CreateResource(User)
        context.renderer = 'json'
        response = Create(context, unit.request).edit_form_get_view()
        assert len(response.json['form']) > 0

    def test_create_user_form(self, unit):
        unit.drop_db()
        unit.create_db()
        context = CreateResource(User)
        post = [
            ('form.submitted', '1'),
            ('__formid__', 'deform'),
            ('__start__', ':mapping'),
            ('id', '1'),
            ('name', 'foo bar baz'),
            ('__end__', ':mapping')
        ]
        unit.request.POST = OrderedDict(post)
        unit.request.params = OrderedDict(post)
        unit.request.sacrud_prefix = 'sacrud'
        context.renderer = 'json'
        response = Create(context, unit.request).edit_form_post_view()
        assert isinstance(response, HTTPFound)
        user = unit.dbsession.query(User).filter_by(id=1).first()
        assert user.id == 1
        assert user.name == 'foo bar baz'


class UpdateTests(object):

    def test_update_user_form(self, unit):
        context = UpdateResource(User)
        context.renderer = 'json'
        response = Create(context, unit.request).edit_form_get_view()
        assert len(response.json['form']) > 0


tests/test_common.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2014 uralbash <root@uralbash.ru>
#
# Distributed under terms of the MIT license.

import colander
import sqlalchemy
from sqlalchemy import Column, String, Integer
from sacrud_deform import SacrudForm
from ps_alchemy.views import preprocessing_value
from sqlalchemy.ext.declarative import declarative_base


def test_preprocessing_value():

    Base = declarative_base()  # noqa

    class User(Base):
        __tablename__ = 'users'
        id = Column(Integer, primary_key=True)
        foo = Column(String)
        bar = Column(Integer)

    form = SacrudForm(None, None, User)
    form.make_appstruct()
    assert preprocessing_value("foo", colander.null, form.schema) == ""
    assert preprocessing_value("foo", "", form.schema) == ""
    assert isinstance(
        sqlalchemy.sql.null(),
        type(preprocessing_value("bar", "", form.schema))
    )
    value = {'foo': 'bar'}
    assert preprocessing_value("foo", value, form.schema) == value
    assert preprocessing_value("foo", u"говядо", form.schema) == u"говядо"
    assert preprocessing_value(
        "foo",
        u'\u0433\u043e\u0432\u044f\u0434\u043e', form.schema) == u"говядо"


ps_alchemy/views.py
import json
import logging

import deform
import colander
import peppercorn
import sqlalchemy
import transaction
from pyramid.view import view_config, view_defaults
from sacrud.common import pk_list_to_dict
from pyramid.compat import escape, text_type
from pyramid_sacrud import PYRAMID_SACRUD_VIEW
from pyramid.renderers import render_to_response
from sqlalchemy.orm.exc import NoResultFound
from paginate_sqlalchemy import SqlalchemyOrmPage
from pyramid.httpexceptions import HTTPFound, HTTPNotFound
from pyramid_sacrud.exceptions import SacrudException
from pyramid_sacrud.localization import _ps

from .paginator import get_paginator
from .resources import (
    ListResource,
    CreateResource,
    DeleteResource,
    MassActionResource
)


def preprocessing_value(key, value, form):
    for groups in form.children:
        for column in groups:
            if column.name == key:
                if isinstance(column.typ, colander.Number):
                    try:
                        float(text_type(value))
                    except ValueError:
                        value = sqlalchemy.sql.null()
                elif value is colander.null:
                    value = ""
                return value


@view_defaults(permission=PYRAMID_SACRUD_VIEW)
class CRUD(object):

    def __init__(self, context, request):
        self.context = context
        self.request = request

    def commit(self):
        try:
            self.context.dbsession.commit()
        except AssertionError:
            transaction.commit()

    def abort(self):
        try:
            self.context.dbsession.rollback()
        except AssertionError:
            transaction.abort()

    def list_view_response(self):
        return HTTPFound(
            location='/' + self.request.sacrud_prefix + '/' +
            self.request.resource_path(
                self.context.get_list_resource(self.context)
            )
        )

    def flash_message(self, message, status="success"):
        if hasattr(self.request, 'session'):
            self.request.session.flash([message, status])


class Read(CRUD):

    @view_config(
        request_method='GET',
        context=ListResource,
        route_name=PYRAMID_SACRUD_VIEW,
    )
    def list_view(self):
        rows = self.context.sacrud.read()
        try:
            paginator_attr = get_paginator(
                self.request, self.context.items_per_page
            )
        except ValueError:
            raise HTTPNotFound
        params = {
            'paginator': SqlalchemyOrmPage(rows, **paginator_attr)
        }
        return render_to_response(
            self.context.renderer, params, request=self.request
        )


class Create(CRUD):

    @view_config(
        request_method='GET',
        context=CreateResource,
        route_name=PYRAMID_SACRUD_VIEW
    )
    def edit_form_get_view(self):
        params = {'form': self.context.form(self.request).render()}
        return render_to_response(
            self.context.renderer, params, request=self.request
        )

    @view_config(
        request_method='POST',
        context=CreateResource,
        route_name=PYRAMID_SACRUD_VIEW
    )
    def edit_form_post_view(self):
        form = self.context.form(self.request)
        params = {'form': form.render()}

        def get_reponse(form=None):
            if form:
                params['form'] = form
            return render_to_response(
                self.context.renderer, params, request=self.request
            )

        if 'form.submitted' in self.request.params:
            controls = self.request.POST.items()
            pstruct = peppercorn.parse(controls)

            # Validate form
            try:
                deserialized = form.validate_pstruct(pstruct).values()
            except deform.ValidationFailure as e:
                return get_reponse(e.render())
            data = {k: preprocessing_value(k, v, form)  # TODO: optimize it
                    for d in deserialized
                    for k, v in d.items()}

            # Update object
            try:
                if self.context.obj:
                    obj = self.context.sacrud._add(self.context.obj, data)
                    flash_action = 'updated'
                else:
                    obj = self.context.sacrud.create(data)
                    flash_action = 'created'
                name = obj.__repr__()
                self.context.dbsession.flush()
            except SacrudException as e:
                self.flash_message(e.message, status=e.status)
                return get_reponse()
            except Exception as e:
                self.abort()
                logging.exception("Something awful happened!")
                raise e

            self.commit()

            # Make response
            self.flash_message(_ps(
                u"You ${action} object of ${name}",
                mapping={'action': flash_action, 'name': escape(name or '')}
            ))
            return self.list_view_response()
        return get_reponse()


class Delete(CRUD):

    @view_config(
        request_method='POST',
        context=DeleteResource,
        route_name=PYRAMID_SACRUD_VIEW
    )
    def delete_view(self):
        if not hasattr(self.context, 'obj'):
            raise HTTPNotFound
        self.context.dbsession.delete(self.context.obj)
        self.commit()
        self.flash_message(
            _ps("You have removed object of ${name}",
                mapping={'name': escape(text_type(self.context.obj) or '')}))
        return self.list_view_response()

    @view_config(
        request_method='POST',
        context=MassActionResource,
        request_param='mass_action=delete',
        route_name=PYRAMID_SACRUD_VIEW
    )
    def mass_delete_view(self):
        items_list = self.request.POST.getall('selected_item')
        primary_keys = [pk_list_to_dict(json.loads(item))
                        for item in items_list]
        objects = self.context.sacrud.read(*primary_keys)
        try:
            if hasattr(objects, 'delete'):
                object_names = [escape(x.__repr__() or '') for x in objects]
                objects.delete()
            else:
                object_names = [escape(objects.__repr__() or ''), ]
                self.request.dbsession.delete(objects)
        except (NoResultFound, KeyError):
            raise HTTPNotFound
        except SacrudException as e:
            self.flash_message(e.message, status=e.status)
        except Exception as e:
            transaction.abort()
            logging.exception("Something awful happened!")
            raise e
        transaction.commit()
        self.flash_message(_ps("You delete the following objects:"))
        [self.flash_message(obj) for obj in object_names]
        return self.list_view_response()


ps_alchemy/__init__.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2015 uralbash <root@uralbash.ru>
#
# Distributed under terms of the MIT license.
from pyramid.events import ApplicationCreated
from pyramid_sacrud.routes import resources_preparing_factory
from sqlalchemy.ext.declarative.api import DeclarativeMeta

from .resources import ListResource


def models_preparing(app):
    """ Wrap all sqlalchemy model in settings.
    """

    def wrapper(resource, parent):
        if isinstance(resource, DeclarativeMeta):
            resource = ListResource(resource)
        if not getattr(resource, '__parent__', None):
            resource.__parent__ = parent
        return resource

    resources_preparing_factory(app, wrapper)


def includeme(config):
    config.add_subscriber(models_preparing, ApplicationCreated)

    config.include('ps_crud')
    config.include('pyramid_jinja2')
    config.add_jinja2_search_path("templates")
    config.scan('.views')

    # deform
    config.include('sacrud_deform')
    config.add_static_view('deform_static', 'deform:static')


ps_alchemy/resources.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2015 uralbash <root@uralbash.ru>
#
# Distributed under terms of the MIT license.

"""
Provide SQLAlchemy resource for pyramid_sacrud.
"""
from zope.interface import implementer
from zope.sqlalchemy import ZopeTransactionExtension
from pyramid.location import lineage
from pyramid.threadlocal import get_current_registry

import sqlalchemy
from sacrud.action import CRUD
from sacrud.common import pk_to_list, pk_list_to_dict, get_attrname_by_colname
from sacrud_deform import SacrudForm
from sqlalchemy.orm import sessionmaker, scoped_session
from pyramid_sacrud.interfaces import ISacrudResource


class BaseResource(object):

    breadcrumb = True

    def __init__(
            self, table, dbsession=None, name=None, **kwargs
    ):
        self.table = table
        self.kwargs = kwargs
        self._dbsession = dbsession
        self.__name__ = name or getattr(self, '__name__', None)

    @property
    def verbose_name(self):
        return getattr(self.table, 'verbose_name', self.table.__tablename__)

    @property
    def _columns(self):
        columns = getattr(self.table, 'sacrud_list_col',
                          self.table.__table__.c)

        class Column(object):
            def __init__(self, column):
                self.column = column

            def value(self, row):
                return getattr(
                    row,
                    get_attrname_by_colname(row, self.column.name)
                )

            @property
            def name(self):
                return getattr(self.column.info, 'verbose_name',
                               self.column.name)

        return list(Column(col) for col in columns)

    @classmethod
    def _get_id(cls, obj, json=True):
        return pk_to_list(obj, json)

    @property
    def ps_crud(self):
        update = self.get_update_resource
        delete = self.get_delete_resource
        create = self.get_create_resource()
        mass_action = self.get_mass_action_resource()
        return {
            "get_id": self._get_id,
            "columns": self._columns,
            "crud": {
                "mass_action": mass_action,
                "create": create,
                "update": update,
                "delete": delete,
            }
        }

    @property
    def dbsession(self):
        if not self._dbsession:
            self._dbsession = self._default_dbsession
        return self._dbsession

    @dbsession.setter
    def dbsession(self, dbsession):
        self._dbsession = dbsession

    @property
    def _default_dbsession(self):
        registry = get_current_registry()
        engine = sqlalchemy.engine_from_config(registry.settings)
        session = scoped_session(
            sessionmaker(extension=ZopeTransactionExtension())
        )
        session.configure(bind=engine)
        return session

    @property
    def sacrud(self):
        return CRUD(self.dbsession, self.table, commit=False)

    def get_mass_action_resource(self):
        resource = MassActionResource(self.table, self.dbsession)
        resource.__parent__ = self
        return resource

    def get_list_resource(self, resource):
        for context in lineage(resource):
            if isinstance(context, ListResource):
                return context

    def get_create_resource(self):
        resource = CreateResource(self.table, self.dbsession)
        resource.__parent__ = self
        return resource

    def get_update_resource(self, obj=None):
        resource = UpdateResource(self.table, dbsession=self.dbsession)
        resource.__parent__ = self
        if obj:
            pk = pk_to_list(obj)
            for key in pk:
                resource = resource[str(key)]
        return resource

    def get_delete_resource(self, obj=None):
        resource = DeleteResource(self.table, dbsession=self.dbsession)
        resource.__parent__ = self.get_list_resource(self)
        if obj:
            pk = pk_to_list(obj)
            for key in pk:
                resource = resource[str(key)]
        return resource


@implementer(ISacrudResource)
class ListResource(BaseResource):

    title = 'Alchemy view'
    renderer = '/ps_alchemy/crud/list.jinja2'

    def __init__(self, table, dbsession=None):
        self.__name__ = table.__tablename__
        super(ListResource, self).__init__(
            table, dbsession=dbsession
        )

    @property
    def items_per_page(self):
        return int(
            get_current_registry().settings.get(
                'ps_alchemy.items_per_page', 5
            )
        )

    def __getitem__(self, name):
        if name == 'create':
            return self.get_create_resource()
        elif name == 'update':
            return self.get_update_resource()
        elif name == 'delete':
            return self.get_delete_resource()
        elif name == 'mass_action':
            return self.get_mass_action_resource()


class CreateResource(BaseResource):

    _obj = None
    _form = None
    title = 'Alchemy create'
    __name__ = 'create'
    renderer = '/ps_alchemy/crud/create.jinja2'

    @property
    def obj(self):
        return self._obj

    @obj.setter
    def obj(self, obj):
        self._obj = obj
        self.form = SacrudForm(
            obj=self._obj, dbsession=self.dbsession, table=self.table
        )

    @property
    def form(self):
        return self._form or SacrudForm(
            obj=self._obj, dbsession=self.dbsession, table=self.table
        )

    @form.setter
    def form(self, form):
        self._form = form


class PrimaryKeyResource(BaseResource):

    breadcrumb = False

    def get_primary_key(self):
        return list(reversed([
            x.__name__ for x in lineage(self)
            if isinstance(x, PrimaryKeyResource) and
            x.__name__ not in ['update', 'delete']
        ]))

    def _getitem(self, resource, name):
        primary_key = self.get_primary_key()
        primary_key.append(name)
        pk_len = len(self.table.__table__.primary_key) * 2
        if len(primary_key) < pk_len:
            resource = self.__class__(self.table, self.dbsession, name)
            resource.__parent__ = self
            return resource
        elif len(primary_key) == pk_len:
            try:
                obj = self.sacrud.read(pk_list_to_dict(primary_key)).one()
            except sqlalchemy.orm.exc.NoResultFound:
                return None
            resource.obj = obj
            resource.__parent__ = self
            return resource


class UpdateResource(PrimaryKeyResource):

    __name__ = 'update'

    def __getitem__(self, name):
        resource = CreateResource(self.table, self.dbsession, name)
        return self._getitem(resource, name)


class DeleteResource(PrimaryKeyResource):

    __name__ = 'delete'

    def __getitem__(self, name):
        resource = DeleteResource(self.table, self.dbsession, name)
        return self._getitem(resource, name)


class MassActionResource(BaseResource):

    __name__ = 'mass_action'


example/ps_alchemy_example.py
import datetime

from pyramid.config import Configurator
from pyramid.session import SignedCookieSessionFactory

from sqlalchemy import (
    Column,
    String,
    Boolean,
    Integer,
    Unicode,
    DateTime,
    ForeignKey
)
from sqlalchemy.orm import backref, relationship, sessionmaker, scoped_session
from ps_alchemy.resources import ListResource
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()
DBSession = scoped_session(sessionmaker())


class Tree(Base):
    __tablename__ = 'tree'

    id = Column(Integer, primary_key=True)
    parent_id = Column(Integer, ForeignKey("tree.id", ondelete="CASCADE"))
    date_added = Column(
        DateTime(timezone=True),
        default=datetime.datetime.now()
    )


class User(Base):
    __tablename__ = 'user'
    id = Column(Integer, primary_key=True, nullable=False)
    age = Column(Integer)
    name = Column(String(30), primary_key=True)

    def __repr__(self):
        return self.name

    def __json__(self, request):
        return {'name': self.name}


class Group(Base):
    __tablename__ = 'groups'
    id = Column(Integer, primary_key=True)
    name = Column(String(30))

    def __repr__(self):
        return self.name


class Good(Base):
    __tablename__ = 'goods'
    id = Column(Integer, primary_key=True)
    name = Column(String(30))
    group_id = Column(Integer, ForeignKey('groups.id'))
    group = relationship('Group', backref='goods')

    visible = Column(Boolean, default=False)
    archive = Column(Boolean)

    def __repr__(self):
        return self.name


class Parent(Base):
    __tablename__ = 'parents'

    slug = Column(Unicode, primary_key=True)
    name = Column(Unicode, nullable=False)


class Child(Base):
    __tablename__ = 'children'

    provider_slug = Column(Unicode, ForeignKey('parents.slug'),
                           primary_key=True)
    slug = Column(Unicode, primary_key=True)

    provider = relationship('Parent', backref=backref('children'))


class GroupName(object):

    def __init__(self, label, name):
        self.name = name
        self.label = label if label else name

    def __repr__(self):
        return self.name


class GoodResource(ListResource):

    items_per_page = 3

models = (
    ('Catalouge', [Group, GoodResource(Good)]),
    (GroupName('Auth system', 'auth'), [User]),
    ('', []),
    ('foo', [Parent, Child, Tree])
)


def sacrud_settings(config):
    config.include('ps_alchemy')
    config.include('pyramid_sacrud', route_prefix='admin')
    config.registry.settings['pyramid_sacrud.models'] = models


def database_settings(config):
    from sqlalchemy import engine_from_config
    settings = config.registry.settings
    engine = engine_from_config(settings, 'sqlalchemy.')
    Base.metadata.bind = engine
    Base.metadata.drop_all()
    Base.metadata.create_all()


def add_fixtures(dbsession):
    for id, user_name in enumerate(('admin', 'moderator', 'user1', 'user2')):
        dbsession.add(User(name=user_name, id=id))
    for group_name in ('Electronics', 'Fashion', 'Home & Garden', 'Motors'):
        group = Group(name=group_name)
        dbsession.add(group)
        if group_name == 'Electronics':
            dbsession.add(Good(name='iPhone', group=group))
            dbsession.add(Good(name='Fridge', group=group))
            dbsession.add(Good(name='YotaPhone', group=group))
        elif group_name == 'Fashion':
            dbsession.add(Good(name='Jeans', group=group))
        elif group_name == 'Home & Garden':
            dbsession.add(Good(name='Rake', group=group))
        elif group_name == 'Motors':
            dbsession.add(Good(name='Chevrolet Cavalier', group=group))
            dbsession.add(Good(name='LADA Granta', group=group))
    dbsession.commit()


def main(global_settings, **settings):
    my_session_factory = SignedCookieSessionFactory('itsaseekreet')
    config = Configurator(
        settings=settings,
        session_factory=my_session_factory,
    )
    config.include(database_settings)
    config.include(sacrud_settings)

    if 'fixtures' in settings:
        add_fixtures(DBSession)
    return config.make_wsgi_app()


if __name__ == '__main__':
    settings = {
        'sqlalchemy.url': 'sqlite:///example.sqlite',
        'fixtures': True,
    }
    app = main({}, **settings)

    try:
        from waitress import serve
        serve(app, host='0.0.0.0', port=6543)
    except ImportError:
        from wsgiref.simple_server import make_server
        server = make_server('0.0.0.0', 6543, app)
        server.serve_forever()


example/pyramid_sacrud_example.py
from sqlalchemy import Column, String, Boolean, Integer, Unicode, ForeignKey
from pyramid.config import Configurator
from pyramid_sacrud import PYRAMID_SACRUD_HOME
from sqlalchemy.orm import backref, relationship, sessionmaker, scoped_session
from pyramid.session import SignedCookieSessionFactory
from pyramid.security import Allow, forget, remember
from pyramid.httpexceptions import HTTPFound
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()
DBSession = scoped_session(sessionmaker())


class Tree(Base):
    __tablename__ = 'tree'

    id = Column(Integer, primary_key=True)
    parent_id = Column(Integer, ForeignKey("tree.id", ondelete="CASCADE"))


class User(Base):
    __tablename__ = 'user'
    id = Column(Integer, primary_key=True)
    age = Column(Integer)
    name = Column(String(30), primary_key=True)

    def __repr__(self):
        return self.name

    def __json__(self, request):
        return {'name': self.name}


class Group(Base):
    __tablename__ = 'groups'
    id = Column(Integer, primary_key=True)
    name = Column(String(30))

    def __repr__(self):
        return self.name


class Good(Base):
    __tablename__ = 'goods'
    id = Column(Integer, primary_key=True)
    name = Column(String(30))
    group_id = Column(Integer, ForeignKey('groups.id'))
    group = relationship('Group', backref='goods')

    visible = Column(Boolean, default=False)
    archive = Column(Boolean)

    def __repr__(self):
        return self.name


class Parent(Base):
    __tablename__ = 'parents'

    slug = Column(Unicode, primary_key=True)
    name = Column(Unicode, nullable=False)


class Child(Base):
    __tablename__ = 'children'

    provider_slug = Column(Unicode, ForeignKey('parents.slug'),
                           primary_key=True)
    slug = Column(Unicode, primary_key=True)

    provider = relationship('Parent', backref=backref('children'))


class GroupName(object):

    def __init__(self, label, name):
        self.name = name
        self.label = label if label else name

    def __repr__(self):
        return self.name


models = (
    ('Catalouge', [Group, Good]),
    (GroupName('Auth system', 'auth'), [User]),
    ('', []),
    ('foo', [Parent, Child, Tree])
)


def sacrud_settings(config):
    config.include('ps_alchemy')
    config.include('pyramid_sacrud', route_prefix='admin')
    config.registry.settings['pyramid_sacrud.models'] = models


def database_settings(config):
    from sqlalchemy import engine_from_config
    settings = config.registry.settings
    engine = engine_from_config(settings, 'sqlalchemy.')
    Base.metadata.bind = engine
    Base.metadata.drop_all()
    Base.metadata.create_all()


def add_fixtures(dbsession):
    for id, user_name in enumerate(('admin', 'moderator', 'user1', 'user2')):
        dbsession.add(User(name=user_name, id=id))
    for group_name in ('Electronics', 'Fashion', 'Home & Garden', 'Motors'):
        group = Group(name=group_name)
        dbsession.add(group)
        if group_name == 'Electronics':
            dbsession.add(Good(name='iPhone', group=group))
            dbsession.add(Good(name='Fridge', group=group))
            dbsession.add(Good(name='YotaPhone', group=group))
        elif group_name == 'Fashion':
            dbsession.add(Good(name='Jeans', group=group))
        elif group_name == 'Home & Garden':
            dbsession.add(Good(name='Rake', group=group))
        elif group_name == 'Motors':
            dbsession.add(Good(name='Chevrolet Cavalier', group=group))
            dbsession.add(Good(name='LADA Granta', group=group))
    dbsession.commit()


class Root(object):
    __acl__ = [(Allow, 'admin', PYRAMID_SACRUD_HOME), ]

    def __init__(self, request):
        self.request = request


def login(request):
    headers = remember(request, 'admin')
    return HTTPFound(location=request.route_url(PYRAMID_SACRUD_HOME),
                     headers=headers)


def logout(request):
    headers = forget(request)
    return HTTPFound(location=request.route_url(PYRAMID_SACRUD_HOME),
                     headers=headers)


def main(global_settings, **settings):
    my_session_factory = SignedCookieSessionFactory('itsaseekreet')
    config = Configurator(
        settings=settings,
        session_factory=my_session_factory,
    )
    config.include(database_settings)
    config.include(sacrud_settings)

    if 'fixtures' in settings:
        add_fixtures(DBSession)
    return config.make_wsgi_app()


if __name__ == '__main__':
    settings = {
        'sqlalchemy.url': 'sqlite:///example.sqlite',
        'fixtures': True,
    }
    app = main({}, **settings)

    try:
        from waitress import serve
        serve(app, host='0.0.0.0', port=6543)
    except ImportError:
        from wsgiref.simple_server import make_server
        server = make_server('0.0.0.0', 6543, app)
        server.serve_forever()
